R/dv8.R

```diff
@@ -0,0 +1,228 @@
+    # Get the layer number (L0, L1, L2, ...)
+    layer_i <- structure[[i]]$name
+
+    # Get the number of modules for the current layer
+    num_modules <- length(structure[[i]]$nested)
+
+    # Iterate through each module
+    for (j in 1: num_modules) {
+      # Get the module number (M0, M1, M2, ...)
+      LM_name <- stringr::str_split(structure[[i]]$nested[[j]]$name, "/")
+      module_j <- LM_name[[1]][2]
+
+      # Get number of files for the layer/module pair
+      num_files_per_LM <- length(structure[[i]]$nested[[j]]$nested)
+
+      # Iterate through each file
+      for (k in 1: num_files_per_LM) {
+        # Get the file name
+        filename_k <- structure[[i]]$nested[[j]]$nested[[k]]$name
+
+        # Represents a row in the table with the layer #, module #, and filename
+        row_ijk <- list(filename_k, module_j, layer_i)
+
+        # Append current row (made up of the layer number, module number, & filename) to a list
+        cluster_list <- append(cluster_list, list(row_ijk))
+      }
+    }
+  }
+
+  # Create table from list of layer,module,file rows
+  cluster_parsed <- data.table::rbindlist(cluster_list)
+
+  # Rename result columns
+  cluster_parsed <- data.table::setnames(cluster_parsed, c("V1", "V2", "V3"), c("filename", "module", "layer"))
+  # Return parsed cluster data table
+  return(cluster_parsed)
+}
+
+#' Convert dependencies table to a sdsm.json
+#'
+#' Converts table of dependencies from Depends into an *-sdsm.json binary.
+#'
+#' @param depends_table parsed dependencies table
+#' @param project_name project name for the sdsm.json
+#' @param is_sorted whether the json is sorted by src and dest
+#' @export
+#' @family dv8
+dependencies_to_sdsmj <- function(depends_table, project_name, is_sorted=FALSE){
+
+  # Get the files in the src and dest columns
+  depends_files_src <- unique(list(depends_table$src)[[1]])
+  depends_files_dest <- unique(list(depends_table$dest)[[1]])
+  depends_files <- c(depends_files_src, depends_files_dest)
+
+  # Get unique file names
+  depends_files <- unique(depends_files)
+
+  # Sort the file names
+  variables <- sort(depends_files, method="radix")
+
+  # Make indices for the filenames
+  variables_indices <- 1:length(variables)
+  # Make named list with filenames as keys and indices as values
+  names(variables_indices) <- variables
+
+  # List to hold cells' indices
+  cells_indices <- 1:nrow(depends_table)
+
+
+  # get the cell from each row in the depends table
+  getCell <- function(i){
+    depends_src <- depends_table[["src"]][[i]]
+    depends_dest <- depends_table[["dest"]][[i]]
+
+    # Look at first value in parse_depends table, get matching index in vars array
+    src_index <- variables_indices[depends_src] - 1
+    dest_index <- variables_indices[depends_dest] - 1
+
+    values = list()
+    parameters = list("Call", "Import", "Use", "Parameter", "Contain", "Create", "Extend",
+                      "Return","Implement", "Cast", "Throw", "Annotation")
```

Comment:
Does your function assume all types of parameters are always present? Or would it work if a user, for example, subset the output of parse_dependencies() only to contain a subset of them and then provides to your function?

It is very important the latter works, as this would enable full customization of the analysis for dv8 should the user wish to remove types of dependencies in their analysis, and the reason why this is included on config files:

https://github.com/sailuh/kaiaulu/blob/716a46e546d5293f3b2856cace62311c95aae977/conf/thrift.yml#L101-L115

To be clear, the subsetting is done at Notebook level, instead of function level. I just want to make sure your function does not presume all dependencies are always present (which is also impossible to guarantee depending on the project being parsed).

Reply:
Hi Carlos, my function does not assume all parameters are present.
>Or would it work if a user, for example, subset the output of parse_dependencies() only to contain a subset of them and then provides to your function?
Yeah that should work fine. I think for most cells in my json they don't have all dependencies anyway.

Also, referring to my dependencies_to_sdsmj()
```r
    # Get all the parameter values for a specific cell
    for (j in 1: length(parameters)) {
      current_param <- depends_table[i][[parameters[[j]]]]
      if (current_param > 0){
        values[[parameters[[j]]]] <- current_param
      }
    }
```

In the if statement, I check that the parameter I am currently looking at for a particular cell exists, and only then do I add it to the values for a cell.

Reply:
Then it works as intended for me! :)

Reply:
Just as a reminder from our last meeting: We want to make sure your function here can account for parse_dependencies() having less columns (as every column represents a type in the configuration file) should the user decide to include less types:

https://github.com/sailuh/kaiaulu/blob/d73c603a2849b5749e60ea42c977c925a489e1b0/conf/apr.yml#L50-L63

You can see an example of the table here as a reminder: http://itm0.shidler.hawaii.edu/kaiaulu/articles/depends_showcase.html#example-network-visualization-with-igraph-1

Otherwise, there will be no way to leverage the user choice when passing it to DV8. Let me know if this doesn't make sense.